File: modules/graphs/verify_solutions.py
"""
Graph Solution Verifiers  (CSC310 - Graphs)
===========================================

These functions CHECK whether a claimed answer is correct; they do NOT compute
the answer.  Useful as a self-check: run your traversal / MST / shortest-path
output through these.

Includes:
  - verify_topological_order : every directed edge points forward in the order
  - verify_shortest_paths    : distances are feasible (no edge can relax) and,
                               if predecessors are given, each is achieved by a
                               tight tree edge from the source
  - verify_mst               : claimed edges form a spanning tree AND are minimum
                               (checked via the cycle property)
"""
from collections import defaultdict


# ---------- Topological order ----------
def verify_topological_order(n, directed_edges, order):
    if sorted(order) != list(range(n)):
        return False, "order is not a permutation of all vertices"
    pos = {v: i for i, v in enumerate(order)}
    for u, v in directed_edges:
        if pos[u] >= pos[v]:
            return False, f"edge {u}->{v} points backward (pos {pos[u]} >= {pos[v]})"
    return True, "valid topological order"


# ---------- Shortest paths ----------
def verify_shortest_paths(edges, source, dist, pred=None):
    if dist.get(source, None) != 0:
        return False, f"dist[source]={dist.get(source)} (should be 0)"
    # Feasibility: no edge may still be relaxed (triangle inequality is tight).
    for u, v, w in edges:
        if dist[u] + w < dist[v]:
            return False, f"edge {u}->{v} (w={w}) can still relax: {dist[u]}+{w} < {dist[v]}"
    # Achievability: each non-source vertex's distance is realized by a tight
    # predecessor edge leading back toward the source.
    if pred is not None:
        adj = {(u, v): w for u, v, w in edges}
        for v in dist:
            if v == source:
                continue
            p = pred.get(v)
            if p is None:
                return False, f"vertex {v} has no predecessor"
            w = adj.get((p, v))
            if w is None or dist[p] + w != dist[v]:
                return False, f"dist[{v}]={dist[v]} not achieved by tight edge {p}->{v}"
    return True, "distances are correct shortest-path values"


# ---------- Minimum spanning tree ----------
def _tree_path_max(tree_adj, start, goal):
    """Heaviest edge weight on the unique path start..goal in a tree (or None)."""
    stack = [(start, None, -1)]
    visited = {start}
    parent = {start: (None, -1)}
    while stack:
        node, _, _ = stack.pop()
        if node == goal:
            break
        for nb, w in tree_adj[node]:
            if nb not in visited:
                visited.add(nb)
                parent[nb] = (node, w)
                stack.append((nb, node, w))
    if goal not in parent:
        return None
    cur, best = goal, -1
    while parent[cur][0] is not None:
        prev, w = parent[cur]
        best = max(best, w)
        cur = prev
    return best


def verify_mst(n, all_edges, mst_edges):
    if len(mst_edges) != n - 1:
        return False, f"a spanning tree of {n} vertices needs {n-1} edges, got {len(mst_edges)}"
    # spanning + acyclic via union-find
    parent = list(range(n))

    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    for u, v, w in mst_edges:
        ru, rv = find(u), find(v)
        if ru == rv:
            return False, f"edge {u}-{v} creates a cycle (not a tree)"
        parent[ru] = rv
    roots = {find(x) for x in range(n)}
    if len(roots) != 1:
        return False, "claimed tree does not connect all vertices"

    total = sum(w for _, _, w in mst_edges)

    # minimality via the cycle property: every non-tree edge must be at least as
    # heavy as the heaviest edge on the tree path between its endpoints.
    tree_adj = defaultdict(list)
    tree_set = {frozenset((u, v)) for u, v, _ in mst_edges}
    for u, v, w in mst_edges:
        tree_adj[u].append((v, w))
        tree_adj[v].append((u, w))
    for u, v, w in all_edges:
        if frozenset((u, v)) in tree_set:
            continue
        heaviest = _tree_path_max(tree_adj, u, v)
        if heaviest is not None and w < heaviest:
            return False, (f"not minimal: non-tree edge {u}-{v} (w={w}) is lighter "
                           f"than tree-path max {heaviest}")
    return True, f"valid minimum spanning tree, total weight {total}"


def _report(title, result):
    ok, msg = result
    print(f"{title}: {'PASS' if ok else 'FAIL'} - {msg}")


def main():
    print("== Topological order ==")
    dag = [(0, 1), (0, 2), (1, 3), (2, 3), (3, 4)]
    _report("correct order [0,1,2,3,4]", verify_topological_order(5, dag, [0, 1, 2, 3, 4]))
    _report("bad order     [1,0,2,3,4]", verify_topological_order(5, dag, [1, 0, 2, 3, 4]))

    print("\n== Shortest paths (source 0) ==")
    sp = [(0, 1, 1), (0, 2, 4), (1, 2, 2), (1, 3, 6), (2, 3, 3), (3, 4, 1)]
    good = {0: 0, 1: 1, 2: 3, 3: 6, 4: 7}
    bad = {0: 0, 1: 1, 2: 4, 3: 6, 4: 7}   # dist[2] too big -> edge 1->2 relaxes
    _report("correct distances", verify_shortest_paths(sp, 0, good))
    _report("bad distances    ", verify_shortest_paths(sp, 0, bad))

    print("\n== Minimum spanning tree ==")
    g = [(0, 1, 2), (0, 2, 3), (1, 2, 1), (1, 3, 4), (2, 3, 5), (3, 4, 6)]
    good_mst = [(1, 2, 1), (0, 1, 2), (1, 3, 4), (3, 4, 6)]    # weight 13
    bad_mst = [(1, 2, 1), (0, 2, 3), (1, 3, 4), (3, 4, 6)]     # spanning but weight 14
    _report("correct MST (w=13)", verify_mst(5, g, good_mst))
    _report("non-minimal tree  ", verify_mst(5, g, bad_mst))


if __name__ == "__main__":
    main()

File: modules/graphs/verify_solutions.cpp
// Graph Solution Verifiers  (CSC310 - Graphs)
// ===========================================
//
// UTILITY (safe).  These functions CHECK whether a claimed answer is correct;
// they do NOT compute it.  A verifier cannot be submitted as a Project 3 or lab
// solution -- it validates output produced some other way.  Port of
// verify_solutions.py.
//
//   verify_topological_order : every directed edge points forward in the order
//   verify_shortest_paths    : no edge can still relax (distances are feasible)
//   verify_mst               : claimed edges form a spanning tree AND are minimum
//                              (checked via the cycle property)
//
// Build:  g++ -std=c++17 -O2 verify_solutions.cpp -o verify_solutions && ./verify_solutions
#include <algorithm>
#include <functional>
#include <iostream>
#include <map>
#include <set>
#include <string>
#include <tuple>
#include <vector>

using std::pair;
using std::string;
using std::vector;
using Edge = std::tuple<int, int, int>;

pair<bool, string> verifyTopo(int n, const vector<pair<int, int>>& edges,
                              const vector<int>& order) {
    vector<int> sorted_order = order;
    std::sort(sorted_order.begin(), sorted_order.end());
    for (int i = 0; i < n; ++i)
        if (sorted_order[i] != i) return {false, "order is not a permutation of all vertices"};
    vector<int> pos(n);
    for (int i = 0; i < (int)order.size(); ++i) pos[order[i]] = i;
    for (auto& [u, v] : edges)
        if (pos[u] >= pos[v])
            return {false, "edge " + std::to_string(u) + "->" + std::to_string(v) + " points backward"};
    return {true, "valid topological order"};
}

pair<bool, string> verifyShortestPaths(const vector<Edge>& edges, int source,
                                       const std::map<int, int>& dist) {
    auto it = dist.find(source);
    if (it == dist.end() || it->second != 0)
        return {false, "dist[source] should be 0"};
    for (auto& [u, v, w] : edges) {
        if (dist.at(u) + w < dist.at(v))
            return {false, "edge " + std::to_string(u) + "->" + std::to_string(v) +
                           " (w=" + std::to_string(w) + ") can still relax"};
    }
    return {true, "distances are correct shortest-path values"};
}

// heaviest edge weight on the unique tree path a..b (or -1 if unreachable)
int treePathMax(const std::map<int, vector<pair<int, int>>>& adj, int a, int b) {
    std::map<int, pair<int, int>> parent; // node -> (prev, weight)
    std::set<int> visited{a};
    parent[a] = {-1, -1};
    vector<int> stack{a};
    while (!stack.empty()) {
        int node = stack.back();
        stack.pop_back();
        if (node == b) break;
        auto it = adj.find(node);
        if (it == adj.end()) continue;
        for (auto& [nb, w] : it->second)
            if (!visited.count(nb)) {
                visited.insert(nb);
                parent[nb] = {node, w};
                stack.push_back(nb);
            }
    }
    if (!parent.count(b)) return -1;
    int cur = b, best = -1;
    while (parent[cur].first != -1) {
        best = std::max(best, parent[cur].second);
        cur = parent[cur].first;
    }
    return best;
}

pair<bool, string> verifyMST(int n, const vector<Edge>& all_edges,
                             const vector<Edge>& mst) {
    if ((int)mst.size() != n - 1)
        return {false, "spanning tree needs " + std::to_string(n - 1) + " edges"};
    vector<int> parent(n);
    for (int i = 0; i < n; ++i) parent[i] = i;
    std::function<int(int)> find = [&](int x) {
        while (parent[x] != x) { parent[x] = parent[parent[x]]; x = parent[x]; }
        return x;
    };
    for (auto& [u, v, w] : mst) {
        int ru = find(u), rv = find(v);
        if (ru == rv) return {false, "claimed tree contains a cycle"};
        parent[ru] = rv;
    }
    std::set<int> roots;
    for (int i = 0; i < n; ++i) roots.insert(find(i));
    if (roots.size() != 1) return {false, "claimed tree does not connect all vertices"};

    int total = 0;
    std::map<int, vector<pair<int, int>>> tadj;
    std::set<pair<int, int>> tset;
    for (auto& [u, v, w] : mst) {
        total += w;
        tadj[u].push_back({v, w});
        tadj[v].push_back({u, w});
        tset.insert({std::min(u, v), std::max(u, v)});
    }
    for (auto& [u, v, w] : all_edges) {
        if (tset.count({std::min(u, v), std::max(u, v)})) continue;
        int heaviest = treePathMax(tadj, u, v);
        if (heaviest != -1 && w < heaviest)
            return {false, "not minimal: non-tree edge " + std::to_string(u) + "-" +
                           std::to_string(v) + " (w=" + std::to_string(w) +
                           ") is lighter than tree-path max " + std::to_string(heaviest)};
    }
    return {true, "valid minimum spanning tree, total weight " + std::to_string(total)};
}

void report(const string& title, const pair<bool, string>& r) {
    std::cout << title << ": " << (r.first ? "PASS" : "FAIL") << " - " << r.second << "\n";
}

int main() {
    std::cout << "== Topological order ==\n";
    vector<pair<int, int>> dag = {{0, 1}, {0, 2}, {1, 3}, {2, 3}, {3, 4}};
    report("correct order [0,1,2,3,4]", verifyTopo(5, dag, {0, 1, 2, 3, 4}));
    report("bad order     [1,0,2,3,4]", verifyTopo(5, dag, {1, 0, 2, 3, 4}));

    std::cout << "\n== Shortest paths (source 0) ==\n";
    vector<Edge> sp = {{0,1,1},{0,2,4},{1,2,2},{1,3,6},{2,3,3},{3,4,1}};
    report("correct distances", verifyShortestPaths(sp, 0, {{0,0},{1,1},{2,3},{3,6},{4,7}}));
    report("bad distances    ", verifyShortestPaths(sp, 0, {{0,0},{1,1},{2,4},{3,6},{4,7}}));

    std::cout << "\n== Minimum spanning tree ==\n";
    vector<Edge> g = {{0,1,2},{0,2,3},{1,2,1},{1,3,4},{2,3,5},{3,4,6}};
    report("correct MST (w=13)", verifyMST(5, g, {{1,2,1},{0,1,2},{1,3,4},{3,4,6}}));
    report("non-minimal tree  ", verifyMST(5, g, {{1,2,1},{0,2,3},{1,3,4},{3,4,6}}));
    return 0;
}
